nomt/src/page_region.rs

```diff
@@ -0,0 +1,259 @@
+//! Implements a regioning scheme for the page tree for use as a [`crate::rw_pass_cell::Region`].
+//! This enables safe, sharded, mutable access to portions of the page tree.
+//!
+//! Each [`PageRegion`] spans an entire sub-section of the page tree for exclusive access.
+//! This is either a page P and all of its possible descendants, or an inclusive range of a
+//! page P's child pages and all of their descendants.
+//!
+//! In the first case, all of the P's ancestors are considered as non-exclusive, and in the second
+//! case P and all of its ancestors are considered as non-exclusive.
+//!
+//! That is, a region either implies this:
+//!
+//! ```ignore
+//!                             ┌─┐
+//!                             └┼┘
+//!                              │
+//!            Shared Here & ── ┌▼┐
+//!            Above            └┬┘
+//!                              │
+//!                             ┌▼┐ Exclusive Here & Below
+//!                          ┌──┴┬┴──┐
+//!                          │   │   │
+//!                         ┌▼┐ ┌▼┐ ┌▼┐
+//!                         └─┘ └─┘ └─┘
+//! ```
+//!
+//! or this:
+//!
+//! ```ignore
+//!                             ┌─┐
+//!                             └┼┘
+//!                              │
+//!                             ┌▼┐   Shared Here &
+//!                    ┌───┬───┬┴─┘   Above
+//!                    │   │   │
+//! Exclusive Here &  ┌▼┐ ┌▼┐ ┌▼┐ │ ┌─┐
+//! Below             └─┘ └─┘ └─┘ │ └─┘
+//!                               │
+//! ```
+
+use nomt_core::page_id::{ChildPageIndex, PageId, ROOT_PAGE_ID};
+
+use crate::rw_pass_cell::{Region, RegionContains};
+
+/// A region of pages for shared/exclusive access.
+///
+/// See module docs for more details.
+#[derive(Debug, Clone)]
+pub struct PageRegion {
+    // non-exclusive access: if min is `None` any page along the path to this one (excluding this).
+    //                       otherwise, including this page.
+    // exclusive access: if min is `None`, just to this page ID. otherwise none.
```

Comment:
shouldn't it be: `exclusive access: if min is None, from this page ID. otherwise from min`

Reply:
ah, this comment was meant to just be on this variable, but I will edit for clarity